### PR TITLE
Support -fsanitizer=address on ztest

### PR DESCRIPTION
**Pull request by DeHackEd**
Requires minimum of gcc version 4.8. Only available to ztest
right now, but easily added to other user applications. For
kernel support use the Linux KASAN feature instead.

Signed-off-by: DHE <git@dehacked.net>

### Description
If your compiler supports `-fsanitize=address` and you configure with `--enable-debug` most libraries, `zdb` and `ztest` will be built with it enabled.

### Motivation and Context
By request from @behlendorf in #6877 

### How Has This Been Tested?
Build testing for the disabled path, full test of executable launch abilities for the debug path.

### Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [X] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [X] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by DeHackEd** [comment]
Surprises:

* `-fsanitize=address` adds a lot to the stack sizes of functions. When building with ASAN I raised the stack size from 1024 to 3072. The highest stack size I noticed when building was 2832.
* To be useful, both ztest and libzpool at a minimum would require ASAN enabled. In the end it just made sense to enable it for all the libraries and to `LDFLAGS` for binaries in general.

In a little while I will push a revert to #6877 to see it in action on the buildbots. :)

**Comment by DeHackEd** [comment]
Buildbots failing becuase libasan is not installed by defualt. On CentOS 7 `libasan` needs to be installed. :/

**Comment by behlendorf** [comment]
@DeHackEd thanks!  You can add `libasan` to the installed dependencies by opening a PR to update this script and tagging @dinatale2 for a review to get it merged.

 https://github.com/zfsonlinux/zfs-buildbot/blob/master/scripts/bb-dependencies.sh

**Comment by dinatale2** [comment]
FYI @DeHackEd zfsonlinux/zfs-buildbot#123 was just merged.

**Comment by DeHackEd** [comment]
Thanks. I've pushed a rebase and fix for minor issues in the PR to kick off the buildbots again.

**Comment by DeHackEd** [comment]
Update pushed. This version tries a full link in the configure script to detect systems where `libasan` is missing or broken and doesn't prevent building.

File: module/zfs/spa.c [hunk 1]
```diff
@@ -1647,7 +1647,13 @@ spa_load_l2cache(spa_t *spa)
 	VERIFY(nvlist_remove(sav->sav_config, ZPOOL_CONFIG_L2CACHE,
 	    DATA_TYPE_NVLIST_ARRAY) == 0);
 
-	l2cache = kmem_alloc(sav->sav_count * sizeof (void *), KM_SLEEP);
+	/*
+	 * -fsanitize in userspace will detect kmem_alloc(0, ...)
+	 * as a leak
+	 */
```
Comment (behlendorf) [inline module/zfs/spa.c]: No comment needed, it is a leak and should be fixed.

File: module/zfs/spa.c [hunk 2]
```diff
@@ -1647,7 +1647,13 @@ spa_load_l2cache(spa_t *spa)
...
+	if (sav->sav_count > 0)
+		l2cache = kmem_alloc(sav->sav_count * sizeof (void *),
```
Comment (behlendorf) [inline module/zfs/spa.c]: It looks like this same issue can occur on line 1588, I'm surprised it wasn't also flagged.

File: cmd/ztest/Makefile.am [hunk 3]
```diff
@@ -2,7 +2,7 @@ include $(top_srcdir)/config/Rules.am
 
 # -Wnoformat-truncation to get rid of compiler warning for unchecked
 #  truncating snprintfs on gcc 7.1.1.
-AM_CFLAGS += $(DEBUG_STACKFLAGS) $(FRAME_LARGER_THAN) $(NO_FORMAT_TRUNCATION)
+AM_CFLAGS += $(DEBUG_STACKFLAGS) $(FRAME_LARGER_THAN) $(NO_FORMAT_TRUNCATION) $(COMPILER_FSANITIZE_ADDRESS)
```
Comment (behlendorf) [inline cmd/ztest/Makefile.am]: Let's pull `$(COMPILER_FSANITIZE_ADDRESS)` down on to it's own line to mind the 80 character limit.  Here and elsewhere.

```
AM_CFLAGS += $(DEBUG_STACKFLAGS) $(FRAME_LARGER_THAN) $(NO_FORMAT_TRUNCATION)
AM_CFLAGS += $(COMPILER_FSANITIZE_ADDRESS)
```

File: config/user-compiler-options.m4 [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+		fi
+		AC_MSG_RESULT([yes])
+	],
+	[
+		FRAME_LARGER_THAN=
+		AC_MSG_RESULT([no])
+	])
+
+	CFLAGS="$saved_flags"
+        AC_SUBST([FRAME_LARGER_THAN])
```
Comment (behlendorf) [inline config/user-compiler-options.m4]: No objection to consolidating these, but please fix the whitespace on this line while you're here.  Let's also pull in `user-no-format-truncation.m4`.

File: config/user-compiler-options.m4 [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+dnl #
+AC_DEFUN([ZFS_AC_CONFIG_USER_FRAME_LARGER_THAN], [
+	AC_MSG_CHECKING([for -Wframe-larger-than=<size> support])
+
+	saved_flags="$CFLAGS"
+	CFLAGS="$CFLAGS -Wframe-larger-than=1024"
+
+	AC_COMPILE_IFELSE([AC_LANG_PROGRAM([], [])],
+	[
+		if test -z "$COMPILER_FSANITIZE_ADDRESS"
```
Comment (behlendorf) [inline config/user-compiler-options.m4]: The `AS_IF` macro if available for the conditional.  However, I think it'd be reasonable to simply bump this unconditionally to 4096 to cover both cases.  This only applies to user space where we have headroom, and this was originally added before the kernel build system set `-Wframe-larger-than` by default.

File: config/user-compiler-options.m4 [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+			AC_MSG_RESULT(yes)
+		],[
+			COMPILER_FSANITIZE_ADDRESS=""
+ 			AC_MSG_RESULT(no)
+		])
+
+		CFLAGS="$saved_flags"
+		LDFLAGS="$saved_ldflags"
+	fi
+	AC_SUBST([COMPILER_FSANITIZE_ADDRESS])
```
Comment (behlendorf) [inline config/user-compiler-options.m4]: I'd suggest either dropping the `COMPILER_` prefix, which isn't super useful and would be my preference.  Or alternately adding it to NO_FORMAT_TRUNCATION and FRAME_LARGER_THAN for consistency.

File: config/zfs-build.m4 [hunk 7]
```diff
@@ -27,7 +27,7 @@ AC_DEFUN([ZFS_AC_DEBUG], [
 	AC_MSG_CHECKING([whether assertion support will be enabled])
 	AC_ARG_ENABLE([debug],
 		[AS_HELP_STRING([--enable-debug],
-		[Enable assertion support @<:@default=no@:>@])],
+		[Enable assertion support and compiler features where available @<:@default=no@:>@])],
```
Comment (behlendorf) [inline config/zfs-build.m4]: The way you've implemented this setting `--enable-debug` has no effect of this compiler feature.  It only defines DEBUG and sets `-Werror`, so I'm assuming this is an unrelated bit of clarification.  In which case, how about the following message instead.

```
Enable assertions and -Werror @<:@default=no@:>@
```

Comment (DeHackEd) [reply config/zfs-build.m4]: Actually it does. `-fsanitize` is only checked for and hence enabled when a debug build is performed. The whole new m4 block is guarded with
```m4
if test "$enable_debug" = "yes"
```
This allows a normal build to operate without `-fsanitize` even on systems which support it.

But I agree, this text needs rewording.

Comment (behlendorf) [reply config/zfs-build.m4]: Ahh, I see.  I glossed over that in my initial reading.  What do you think about moving the call points for these debug-only compiler checks in to `ZFS_AC_DEBUG_ENABLE`.  This way these M4 macro's will only run for debug builds where they're needed.

Comment (DeHackEd) [reply config/zfs-build.m4]: I wanted to keep the stack size check together with this one for the sake of coherence and so that non-debug builds would still use a sane stack size check value. m4 is new to me so I'm sure my style is horrible.

**Comment by DeHackEd** [comment]
Regarding the memory leak, the `l2cache` variable is stack-local whereas `newvdevs` will be assigned to `sav->sav_vdevs` to be freed later in `spa_unload()`.

That said I have been hitting another leak in the refcount code but I'm not getting much information about where the allocation is coming from yet. We still need that fixed before we push this branch.

**Comment by behlendorf** [comment]
Closing, replaced by #6941
